Add how-to to configure TLS to access the 12-Factor App over HTTPS

javierdelapuente opened the pull request:
Add a how-to (nested under the ingress) to explain how to access your 12-factor over HTTPS.

javierdelapuente commented:
@erinecon can you have a first look at this? Thanks!

erinecon commented on docs/howto/manage-web-app-charms/integrate-web-app-charm.rst:
```diff
@@ -89,6 +89,54 @@ project file.
 .. _integrate_web_app_cos:
 
+
+Configure TLS to Access the 12-Factor App over HTTPS
```
```suggestion
Configure TLS to access the 12-Factor app over HTTPS
```

erinecon commented on docs/howto/manage-web-app-charms/integrate-web-app-charm.rst:
I would move the new how-to into `docs/howto/manage-web-app-charms/use-web-app-charm.rst` -- it feels like a more appropriate place for this guide. What do you think?

javierdelapuente replied:
Not sure, I was going for there initially, but the TLS is really related to ingress (and not so much about pure 12-factor).

erinecon replied:
Hmmm, good point that this content isn't strictly related to 12-factor... putting that aside for now, I do think that this section should be moved to the "Use" page. Configuring TLS is much less about *integrating* with ingress and more about *using* ingress to me.

erinecon commented on docs/howto/manage-web-app-charms/integrate-web-app-charm.rst:
```diff
@@ -89,6 +89,54 @@ project file.
+
+For example, using the
+`Nginx Ingress Integrator <https://charmhub.io/nginx-ingress-integrator>`_
+charm as an ingress provider charm, you can set the hostname with:
```
```suggestion
charm as an ingress provider, you can set the hostname with:
```
Nitpick, trying to remove the double "charm" mentions

erinecon commented on docs/howto/manage-web-app-charms/integrate-web-app-charm.rst:
```diff
@@ -89,6 +89,54 @@ project file.
+    juju config nginx-ingress-integrator service-hostname=<yourdomain.example.com>
+
+Deploy the ``self-signed-certificates`` charm and integrate with the
+``nginx-ingress-integrator`` app:
```
```suggestion
``nginx-ingress-integrator`` charm:
```

erinecon commented on docs/howto/manage-web-app-charms/integrate-web-app-charm.rst:
```diff
@@ -89,6 +89,54 @@ project file.
+    juju deploy self-signed-certificates
+    juju integrate self-signed-certificates nginx-ingress-integrator
+
+At this point your 12 factor application can be accessed using HTTPS.
```
```suggestion
At this point your 12-factor app can be accessed using HTTPS.
```

erinecon commented on docs/howto/manage-web-app-charms/integrate-web-app-charm.rst:
```diff
@@ -89,6 +89,54 @@ project file.
+HTTP status code that will redirect to the https URL. You can access the HTTPS
+URL of your app with a command like:
+
+.. code-block:: bash
+
+   curl -v --insecure https://<yourdomain.example.com> --resolve <yourdomain.example.com>:443:<ingress-ip>
```
```suggestion
.. code-block:: bash

   curl -v --insecure https://<yourdomain.example.com> \
     --resolve <yourdomain.example.com>:443:<ingress-ip>
```
Nitpick, I think the command is a bit more readable in the rendered output when split over two lines

erinecon commented on docs/howto/manage-web-app-charms/integrate-web-app-charm.rst:
```diff
@@ -89,6 +89,54 @@ project file.
+
+.. note::
+
+    The ``--insecure`` option is needed because the CA in the
```
Nitpick, but you don't define CA until after this note. I'm leaning toward writing out "certificate authority" in the note -- what do you think?

erinecon reviewed (commented):
Thanks so much for writing this how-to guide! Most of my comments are nitpicks, with one exception: I think that the guide shouldn't be under the "Integrate" section.

medubelko commented on docs/howto/manage-web-app-charms/use-web-app-charm.rst:
```diff
@@ -347,3 +347,52 @@ The Pebble logs are available via Grafana or Loki and can be viewed in
+---------------------------------
+
+You can configure TLS if your ingress provider charm supports integration with
+a certificate provider charm. To test locally, you can use the
```
> To test...

By saying "test" here in the first paragraph, we're making what looks like another task statement that contradicts the heading. Is the user going to _configure_ TLS, or _test_ with a charm locally?

Instead, I think we ought to say that the best way to test the TLS connection is with that charm.

medubelko commented on docs/howto/manage-web-app-charms/use-web-app-charm.rst:
```diff
@@ -347,3 +347,52 @@ The Pebble logs are available via Grafana or Loki and can be viewed in
+First, :ref:`integrate your web app with ingress <integrate_web_app_ingress>`.
+
+Using the `Nginx Ingress Integrator <https://charmhub.io/nginx-ingress-integrator>`_
+charm as an ingress provider, you can set the hostname with:
```
```suggestion
charm as an ingress provider, set the hostname with:
```

medubelko commented on docs/howto/manage-web-app-charms/use-web-app-charm.rst:
```diff
@@ -347,3 +347,52 @@ The Pebble logs are available via Grafana or Loki and can be viewed in
+
+    juju config nginx-ingress-integrator service-hostname=<yourdomain.example.com>
+
+Deploy the ``self-signed-certificates`` charm and integrate with the
+``nginx-ingress-integrator`` charm:
```
```suggestion
Deploy the `Self Signed X.509 Certificates
<https://charmhub.io/self-signed-certificates>` charm and integrate the two charms:
```

medubelko commented on docs/howto/manage-web-app-charms/use-web-app-charm.rst:
```diff
@@ -347,3 +347,52 @@ The Pebble logs are available via Grafana or Loki and can be viewed in
+
+At this point your 12-factor app can be accessed using HTTPS.
+In the case of the ``nginx-ingress-integrator`` charm, if you access the
+external URL of your charm using HTTP, you will get a 308 Permanent Redirect
```
> your charm

I think this should be "the integrator charm" or "your app"?

medubelko commented on docs/howto/manage-web-app-charms/use-web-app-charm.rst:
```diff
@@ -347,3 +347,52 @@ The Pebble logs are available via Grafana or Loki and can be viewed in
+    juju deploy self-signed-certificates
+    juju integrate self-signed-certificates nginx-ingress-integrator
+
+At this point your 12-factor app can be accessed using HTTPS.
+In the case of the ``nginx-ingress-integrator`` charm, if you access the
+external URL of your charm using HTTP, you will get a 308 Permanent Redirect
+HTTP status code that will redirect to the HTTPS URL. You can access the HTTPS
+URL of your app with a command like:
```
```suggestion
Your 12-factor app is now accessible over HTTPS. If you access the external URL of your
app using HTTP, it returns an HTTP ``308 Permanent Redirect`` status and a redirect to
the HTTPS URL. You can access the HTTPS URL of your app with a command like:
```

medubelko commented on docs/howto/manage-web-app-charms/use-web-app-charm.rst:
```diff
@@ -347,3 +347,52 @@ The Pebble logs are available via Grafana or Loki and can be viewed in
+
+    The ``--insecure`` option is needed because the certificate authority in the
+    ``self-signed-certificate`` charm is not trusted.
+    The ``--resolve <yourdomain.example.com>:443:<ingress-ip>`` option to the ``curl``
+    command is a way of resolving the hostname of the request without
```
```suggestion
    The ``--resolve <yourdomain.example.com>:443:<ingress-ip>`` option is a way of
    resolving the hostname of the request without
```

medubelko commented on docs/howto/manage-web-app-charms/use-web-app-charm.rst:
```diff
@@ -347,3 +347,52 @@ The Pebble logs are available via Grafana or Loki and can be viewed in
+    setting a DNS record.
+
+To obtain a TLS certificate signed by a trusted certificate authority (CA)
+using the ACME protocol, you can use the  `LEGO <https://charmhub.io/lego>`_
```
```suggestion
using the ACME protocol, use the `LEGO <https://charmhub.io/lego>`_
```

lengau reviewed (approved):
Approving pending the changes requested by Michael. Would it be possible to re-aim this at the `hotfix/3.5` branch? I'll be releasing 3.5.3 soon and if possible it'd be nice for this to go there.

javierdelapuente commented:
> Approving pending the changes requested by Michael. Would it be possible to re-aim this at the `hotfix/3.5` branch? I'll be releasing 3.5.3 soon and if possible it'd be nice for this to go there.

Thanks! Updated to use hotfix/3.5 (and squashed/rebased for it).
